### PR TITLE
Implement `BigFloat`'s rounding modes

### PR DESCRIPTION
**Pull request by HertzDevil**
~~See https://github.com/crystal-lang/crystal/pull/10413#issuecomment-817114473.~~

### CODE REVIEW

**Comment by asterite** [comment]
Isn't it a bit strange that it would raise by default?

Could ties_even be implemented on top of the others by checking if the number is even?

**Comment by asterite** [comment]
For example, with this implementation:

```crystal
  def round(mode : Number::RoundingMode = :ties_even) : self
    case mode
    in .to_zero?
      trunc
    in .to_positive?
      ceil
    in .to_negative?
      floor
    in .ties_away?
      if self >= 0
        ceil
      else
        floor
      end
    in .ties_even?
      if to_big_i.even?
        ceil
      else
        floor
      end
    end
  end
```

And this program:

```crystal
require "big"

puts "Ties even:"
puts "^^^^^^^^^^"
[-3.5, -2.5, -1.5, -0.5, 0.5, 1.5, 2.5, 3.5].each do |float|
  puts "#{float}.round # = #{float.round(:ties_even)}"
  puts "#{float}.to_big_f.round # = #{float.to_big_f.round(:ties_even)}"
  puts "==="
end

puts "Ties away:"
puts "^^^^^^^^^^"
[-3.5, -2.5, -1.5, -0.5, 0.5, 1.5, 2.5, 3.5].each do |float|
  puts "#{float}.round # = #{float.round(:ties_away)}"
  puts "#{float}.to_big_f.round # = #{float.to_big_f.round(:ties_away)}"
end
```

We get the same result in both cases.

The `ties_even` case is not very performant because it has to go through BigInt. However, I prefer it working than raising. We can always improve it later.

Then for `ties_away` it would be nice to have `positive?` and `negative?` functions in the stdlib. Maybe there's a gmp function to compute that efficiently without comparing anything.

**Comment by straight-shoota** [comment]
> Isn't it a bit strange that it would raise by default?

Definitely. If we don't have an implementation, `BigFloat#round` without arguments should error at compile time because it can never work.
But it seems like a good idea to add *an* implementation as the suggested one. Even if it's inefficient.

**Comment by straight-shoota** [comment]
The implementation in https://github.com/crystal-lang/crystal/pull/10618#issuecomment-817148161 is broken for any non-0.5 value.

**Comment by asterite** [comment]
So the special logic is exactly for 0.5 values?

**Comment by straight-shoota** [comment]
Yes. Ties away and ties even means always rounding to the nearest integer, except for the case that the value is exactly between two integers (i.e. at .5). Only then the special rule away or even takes effect.

**Comment by asterite** [comment]
I see. What's the default in Ruby?

**Comment by asterite** [comment]
It seems it's away. Cant we change the default? It's simpler.

**Comment by asterite** [comment]
Nevermind, I guess because we are in 1.0 we can't change it. I personally dislike that default, but it doesn't matter.

**Comment by HertzDevil** [comment]
The correct implementation for `TIES_AWAY` seems to be:

```crystal
if self >= 0
  (self + 0.5).floor
else
  (self - 0.5).ceil
end
```

**Comment by straight-shoota** [comment]
Yes, Ruby's default is ties away.

And we should not just change that. Even without considering backward compatibility. ties_even is the better default, as detailled and agreed upon in #10414

**Comment by asterite** [comment]
Sounds good. Sorry for the noise.

**Comment by straight-shoota** [comment]
This should do for round_even:
```cr
  def round_even
    if self >= 0
      halfway = self + 0.5
    else
      halfway = self - 0.5
    end
    if halfway.integer?
      if halfway.to_i!.even?
        halfway
      else
        halfway - sign
      end
    else
      if self >= 0
        halfway.floor
      else
        halfway.ceil
      end
    end
  end
```

**Review by straight-shoota — APPROVED** [review]

File: src/big/big_float.cr [hunk 1]
```diff
@@ -309,6 +309,43 @@ struct BigFloat < Float
...
+  # rounds away from zero.
+  def round_away : self
+    if self >= 0
+      (self + 0.5).floor
+    else
+      (self - 0.5).ceil
+    end
+  end
+
+  def integer?
```
Comment (asterite) [inline src/big/big_float.cr:345]: Since this is a new public method, could we document it? Otherwise, leave it as private.

**Comment by HertzDevil** [comment]
Do I have to reset `BigFloat.default_precision` after each spec?

**Comment by HertzDevil** [comment]
If float-to-integer conversion is undefined on overflow, we cannot do `halfway.to_i!.even?` whether it's GMP's or our floats, and GMP doesn't have the modulo operator for floats either. So the given definition will produce undefined results for `BigFloat` values outside `Float64`'s representable range.

**Comment by asterite** [comment]
> It might be necessary to do this via to_s instead

Yeah, that's what I was thinking. I think we should do it that way, and improve it later if we find a better way.

There's `mpf_get_d_2exp` but I'm not sure how that could be used here.

**Comment by HertzDevil** [comment]
If `x == d * 2.0 ** exp` where `0.5 <= d.abs < 1`, then (ignoring the special case for `x == 0.5`) `d`'s binary representation must be of the form `0.1.....X1000...`, and we round up if and only if this second-to-last digit `X` is 1. This is how `d_2exp` could be used.

Unfortunately this does not work for `BigFloat`s with a higher precision than `Float64` (53 bits), even for values that don't overflow.

**Comment by HertzDevil** [comment]
Did some basic benchmarking and turns out imitating the modulo 2 call ourselves seems to be faster than calling `mpf_get_str`.

File: spec/std/big/big_float_spec.cr [hunk 2]
```diff
@@ -1,6 +1,13 @@
 require "spec"
 require "big"
 
+private def with_precision(x)
+  old = BigFloat.default_precision
+  BigFloat.default_precision = x
+  yield
+  BigFloat.default_precision = old
+end
```
Comment (Sija) [inline spec/std/big/big_float_spec.cr:9]:
```suggestion
private def with_precision(value)
  prev = BigFloat.default_precision
  begin
    BigFloat.default_precision = value
    yield
  ensure
    BigFloat.default_precision = prev
  end
end
```

**Review by straight-shoota — APPROVED** [review]

File: spec/std/big/big_float_spec.cr [hunk 3]
```diff
@@ -1,6 +1,17 @@
...
 
+private def with_precision(precision)
+  old_precision = BigFloat.default_precision
+  BigFloat.default_precision = precision
+
+  begin
+    yield
+  ensure
+    BigFloat.default_precision = old_precision
+  end
```
Comment (hugopl) [inline spec/std/big/big_float_spec.cr:12]: some lines could be saved :-)
```suggestion
  yield
ensure
  BigFloat.default_precision = old_precision
```

**Comment by hugopl** [comment]
IMO this could be a good one for 1.1 milestone, since this worked on 0.36.1, what's missing for this to get merged? (my comment in the MR is pure cosmetic and can be ignored)

This would close https://github.com/crystal-lang/crystal/issues/10599 , https://github.com/crystal-lang/crystal/issues/5714 and abort https://github.com/crystal-lang/crystal/pull/7126

**Comment by HertzDevil** [comment]
`BigFloat` has nothing to do with `BigDecimal` at all; the latter depends only on `BigInt`.

**Pull request merged**